src/db/transformer.ts

```diff
@@ -13,17 +13,30 @@ import {
   IndexRecord,
 } from './commit';
 import type {Hash} from '../hash';
-import type * as dag from '../dag/mod';
+import * as dag from '../dag/mod';
 import type {ReadonlyJSONValue} from '../json';
 import {HashRefType} from './hash-ref-type';
 import type {Value} from '../kv/store';
 
-export class Transformer {
-  readonly dagWrite: dag.Write;
+export class Transformer<Tx = dag.Write> {
```

Comment:
I think this could be done in a clearer way with a bit more of a type hiearchy

```
abstract class BaseTransformer {
  // no constructor as the base class doesn't take a dag.Read or dag.Write
  protected abstract getChunk(oldHash: OldHash): Promise<dag.Chunk | undefined>;
  protected abstract writeChunk<D extends Value>(
    _oldHash: OldHash,
    newData: D,
    getRefs: (data: D) => readonly NewHash[],
  ): Promise<NewHash>;
  // ... rest is the same as Transformer's impl now
}

abstract class ReadTransformer extends BaseTransformer {
  protected readonly dagRead: dag.Read; 
  constructor(dagRead) {
    super();
    this.dagRead = dagRead;
  }
  protected getChunk(oldHash: OldHash): Promise<dag.Chunk | undefined> {
    return this.dagRead.getChunk(oldHash);
  }
}
class WriteTransformer extends ReadTransformer {
  protected readonly dagWrite: dag.Write;
  constructor(dagWrite: dag.Write) {
    super(dagWrite);
    this.dagWrite = dagWrite;
  }
  protected async writeChunk<D extends Value>(
    _oldHash: OldHash,
    newData: D,
    getRefs: (data: D) => readonly NewHash[],
  ): Promise<NewHash> {
    const {dagWrite} = this;
    const newChunk = dagWrite.createChunk(newData, getRefs(newData));
    await dagWrite.putChunk(newChunk);
    return newChunk.hash;
  }
}

```

Then 
`ComputeHashTransformer extends BaseTransformer`
and 
`FixupTransformer extends WriteTransformer`

and we could also write transformers that just used a `dag.Read` and don't take a write lock (this would work for rewriting GatherVistor as a transformer).

Reply:
Thanks. Will revisit